man/lmi.1
.TH lmi 1 "October 2013" "openlmi-scripts 0.2.2" "User Commands"

.SH NAME
lmi \- meta-command for managing systems with OpenLMI providers
.SH SYNOPSIS
.nf
.B     lmi [options] [(--trace | --notrace)] [-v]... [-h <host>]...
.RS
.B         <command> [<args> ...]
.RE
.B     lmi [options] [(--trace | --notrace)] [-v]... [-h <host>]...
.B     lmi (--help | --version)
.fi
.SH DESCRIPTION
Allows to set a path to file, where messages will be logged. No log file is written at default.
.SH OPTIONS
.TP
.I -c --config-file <config>
Path to a user configuration file. Options specified here override any settings of global configuration file.
.TP
.I -h --host <host>
Hostname of target system.
.TP
.I --hosts-file <hosts>
Path to a file containing target hostnames. Each hostname must be listed on a single line.
.TP
.I --user <user>
Username used in connection to any target host.
.TP
.I --same-credentials
Use the first credentials given for all hosts.
.TP
.I -n --noverify
Do not verify cimom's ssl certificate.
.TP
.I -v
Increase verbosity of output.
.TP
.I --trace
Show tracebacks on errors.
.TP
.I --notrace
Suppress tracebacks for exceptions.
.TP
.I -q --quiet
Supress output except for errors.
.TP
.I --log-file <log_file>
Output file for logging messages.
.TP
.I --namespace <namespace>
Default CIM namespace to use.
.TP
.I -N --no-headings
Don't print table headings.
.TP
.I -H --human-friendly
Print large values in human friendly units (i.e. MB, GB, TB etc.)
.TP
.I -L --lister-format (table | csv)
 Print output of lister commands in CSV or table format. CSV format is more suitable for machine processing. Defaults to table.
.TP
.I --help
Show this text and quit.
.TP
.I --version
Print version of 'lmi' in use and quit.
.SH CONFIGURATION
There is a system-wide configuration file located at
`/etc/openlmi/lmi.conf` and user configuration file searched at `~/.lmirc`.
File path of the later can also be specified with
.B --config-file
option. Options in user configuration overrides system-wide.

The configuration consists of sections, led by a
.B [section]
header and followed by
.B name: value
entries. Comments are prefixed with
.B #
or
.B ;
.PP
Please refer to
.B RFC 822
for more details on this format. Follows a list of possible configuration
options with leading section names.
.TP
.B Main Section
.RS 4
.TP 4
.B CommandNamespace
Default:
.B lmi.scripts.cmd
.IP
Python namespace, where command entry points will be searched for.
.RE
.RS 4
.TP 4
.B Trace
Default:
.B False
.IP
Whether the exceptions should be logged with tracebacks.
.RE
.RS 4
.TP 4
.B Verbosity
Default:
.B 0
.IP
An integer from -1 to 2 saying, how verbosive the output to console shall be. This differs from [Log] Level, which controls the logging messages being logged to file. If logging to console is enabled it sets the minimum severity level. -1 Supresses all messages except for critical ones. 0 shows warnings, 1 info messages and 2 enables debug messages. This option also affects the verbosity of commands, making them print more information to stdout.
.RE
.TP
.B CIM Section
.RS 4
.TP 4
.B namespace
Default:
.B root/cimv2
.IP
To override default CIM namespace, uncomment the line below.
.RE
.TP
.B SSL Section
.RS 4
.TP 4
.B VerifyServerCertificate
Default:
.B True
.IP
Whether to verify server-side certificate, when making secured connection over ssl.
.RE
.TP
.B Format Section
.RS 4
.TP 4
.B HumanFriendly
Default:
.B False
.IP
Whether to print values in human readable forms (e.g. with units).
.RE
.RS 4
.TP 4
.B ListerFormat
Default:
.B table
.IP
What format to use, when listing tabular data. Possible values are: {csv, table}. The former allows for easy machine parsing, the latter is more human friendly.
.RE
.RS 4
.TP 4
.B NoHeadings
Default:
.B False
.IP
Whether to suppress headings (column names) when printing tables.
.RE
.TP
.B Log Section
.RS 4
.TP 4
.B Level
Default:
.B ERROR
.IP
Level can be set to following values: DEBUG, INFO, WARNING, ERROR, CRITICAL Affects only logging to a file.
.RE
.RS 4
.TP 4
.B ConsoleFormat
Default:
.B %(levelname)s: %(message)s
.IP
Format string used, when logging to a console. For details visit: http://docs.python.org/2/library/logging.html#logrecord-attributes
.RE
.RS 4
.TP 4
.B LogToConsole
Default:
.B True
.IP
Whether the logging to console is enabled.
.RE
.RS 4
.TP 4
.B FileFormat
Default:
.B %(asctime)s:%(levelname)-8s:%(name)s:%(lineno)d - %(message)s
.IP
Format string used, when logging to a file.
.RE
.RS 4
.TP 4
.B OutputFile
Allows to set a path to file, where messages will be logged. No log file is written at default.
.RE
.SH AUTHORS
.LP
Michal Minar <miminar@redhat.com>


Changelog
0.2.3
* made localhost a default host
* allow to have a fallback command for command multiplexers
* fixed rendering of instances when PROPERTIES are omitted

0.2.2
* double and single dash can now be used in usage strings
* fixed error handling
* improved logging of errors - handled the same way for all commands
* added [Log] LogToConsole configuration option
* added --notrace and --log-file options
* logging configuration fixes
* support for openlmi-python-base 0.2.1+

0.2.1
* compability fix for newer openlmi-tools
* enhanced check_result() method invocation
* fixed handling of verbosity parameter
* fixed handling of ARG_ARRAY_SUFFIX

0.2.0
* allow embedded credentials in urls of hosts
* allow to override default namespace from config and command line
* added OPT_NO_UNDERSCORES property
* commands api change: replace any sequence of characters matching [^a-z_]
with a single underscore, when renaming options to function arguments
* added LmiInstanceLister base command class
* fixed exit command
* added ARG_ARRAY_SUFFIX
* added --same-credentials flag
* fixed parsing of hosts file

0.1.1
* support for (non)verification of server-side certificates
* allowed *connection* to be an argument in command usage strings
* fixed command declaration checking
* relicensed to BSD
* added namespace property to end-point commands
* associated functions invoked with namespace rather then just connection
object

0.1.0
* initial version with dependency only on docopt


completion/lmi.bash
#
# Copyright (C) 2013 Red Hat, Inc.  All rights reserved.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
# Authors: Roman Rakus <rrakus@redhat.com>
#
# Bash completion for LMI commands

_lmi() {
    local helpers_path="lmi-bash-completion"
    if ! [[ -e "$helpers_path" ]]; then
        helpers_path="/usr/libexec/$helpers_path"
    fi
    local options=(-c --config-file -h --host --hosts-file --user -v --trace -q --quiet -n --noverify --same-credentials --help --version)
    local current="${COMP_WORDS[$COMP_CWORD]}"
    local previous="${COMP_WORDS[COMP_CWORD-1]}"
    local commands=( $($helpers_path/print_possible_commands.sh) )
    local used_command=

    for (( i=1; i < ${#COMP_WORDS[@]} - 1 && i < $COMP_CWORD; i++ )); do
        for (( ci=0; ci < ${#commands[@]}; ci++ )); do
            if [[ ${COMP_WORDS[$i]} == ${commands[$ci]} ]]; then
                used_command=${commands[$ci]}
                break 2;
            fi
        done
    done

    if [[ $used_command ]]; then
        # Check if we have completion executable for command
        if [[ -x $helpers_path/commands/"_$used_command" ]] ; then
            # pass the rest of words typed as parameters
            COMPREPLY=( $($helpers_path/commands/_$used_command "${COMP_WORDS[@]:$((i+1))}") )
        else
            # commands without completion - filename completion
            COMPREPLY=( $(compgen -f -- "$current") )
        fi
    else
        case $current in
            -*) COMPREPLY=( $(compgen "-W ${options[*]}" -- "$current" ) );;
            *) case $previous in
                 -c|--config-file|--hosts-file) COMPREPLY=( $(compgen -f -- "$current" ) );;
                 -h|--host) COMPREPLY=( $(compgen -A hostname -- "$current" ) );;
                 --user) COMPREPLY=( $(compgen -u -- "$current" ) );;
                 --help|--version) ;;
                 *) COMPREPLY=( $(compgen "-W ${commands[*]}" -- "$current" ) );;
               esac
        esac
    fi
}

complete -F _lmi lmi
